presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java

```diff
@@ -1756,6 +1757,83 @@ private void testPartitionPerScanLimit(Session session, HiveStorageFormat storag
         assertFalse(getQueryRunner().tableExists(session, tableName));
     }
 
+    @Test
+    public void testPartitionPerScanLimitSessionProperty()
+    {
+        TestingHiveStorageFormat storageFormat = new TestingHiveStorageFormat(getSession(), HiveStorageFormat.DWRF);
+        testWithStorageFormat(storageFormat, this::testPartitionPerScanLimitSessionProperty);
+    }
+
+    private void testPartitionPerScanLimitSessionProperty(Session session, HiveStorageFormat storageFormat)
+    {
+        Session defaultSession = getSession();
+
+        String tableName = "test_session_partition_per_scan_limit";
+        String partitionsTable = "\"" + tableName + "$partitions\"";
+
+        @Language("SQL") String createTable = "" +
+                "CREATE TABLE " + tableName + " " +
+                "(" +
+                "  foo VARCHAR," +
+                "  part BIGINT" +
+                ") " +
+                "WITH (" +
+                "format = '" + storageFormat + "', " +
+                "partitioned_by = ARRAY[ 'part' ]" +
+                ") ";
+
+        assertUpdate(defaultSession, createTable);
+
+        TableMetadata tableMetadata = getTableMetadata(catalog, TPCH_SCHEMA, tableName);
+        assertEquals(tableMetadata.getMetadata().getProperties().get(STORAGE_FORMAT_PROPERTY), storageFormat);
+        assertEquals(tableMetadata.getMetadata().getProperties().get(PARTITIONED_BY_PROPERTY), ImmutableList.of("part"));
+
+        // insert 500 partitions
+        for (int i = 0; i < 5; i++) {
+            int partStart = i * 100;
+            int partEnd = (i + 1) * 100 - 1;
+
+            @Language("SQL") String insertPartitions = "" +
+                    "INSERT INTO " + tableName + " " +
+                    "SELECT 'bar' foo, part " +
+                    "FROM UNNEST(SEQUENCE(" + partStart + ", " + partEnd + ")) AS TMP(part)";
+
+            assertUpdate(defaultSession, insertPartitions, 100);
+        }
+
+        // create a new session with session partition limit set to no more than 100
+        Session testSession = Session.builder(getSession())
+                .setCatalogSessionProperty(catalog, MAX_PARTITIONS_PER_SCAN, "100")
+                .build();
+
+        // can query less than 100 partitions
+        assertQuery(
+                testSession,
+                "SELECT * FROM " + partitionsTable + " WHERE part >= 100 and part < 110",
+                "VALUES 100, 101, 102, 103, 104, 105, 106, 107, 108, 109");
+        assertQuery(
+                testSession,
+                "SELECT count(foo) FROM " + tableName + " WHERE part < 100",
+                "SELECT 100");
+        // query more than 100 partition will be rejected
+        assertQueryFails(
+                testSession,
+                "SELECT * from " + tableName + " WHERE part < 101",
```

Comment:
Also do these queries exercise the new path you added? At least from the offline discussion, the existing enforcement is not working when you have complex expressions in the filter on partition column.

Reply:
The assertions expect a specific error message, which is newly introduced in the new check. So as long as those error messages are showing up as the assertions expect, the new check is being excised.

Reply:
In what cases the existing implementation in PartitionManager works and doesn't work?